### PR TITLE
etcd operator working group

### PR DESCRIPTION
**Pull request by ahrtr**
Initial PR to create wg-etcd-operator.

/committee steering

Adding stakeholder SIGs

/sig etcd
/sig cluster-lifecycle

cc @jmhbnz @jberkus @serathius @wenjiaswe @fabriziopandini @hakman @neolit123 @justinsb @vincepri

### CODE REVIEW

**Comment by ahrtr** [comment]
We need to finalize:

- liaison
- day/time for the regular meeting
- zoom link for the meeting
- a google doc (archive_url)
- slack channel
- mailing list url

File: sig-list.md
Comment (humblec) [inline sig-list.md:67]: @ahrtr nit:  s/VMWare/VMware ?

File: sigs.yaml
Comment (humblec) [inline sigs.yaml:3386]: s/Note/Note:  ?

File: wg-etcd-operator/charter.md
Comment (humblec) [inline wg-etcd-operator/charter.md:59]: nit: s/proejct/project

File: sig-list.md
Comment (ahrtr) [reply sig-list.md]: The wrong spelling "VMWare" was used in sigs.yaml, so ran into the error [ERROR: wg-etcd-operator: chairs: expected person: {ahrtr Benjamin Wang VMWare}, got: {ahrtr Benjamin Wang VMware}](https://github.com/kubernetes/community/blob/38918f9f128a3c573b54b3d5ae9cb0ab273c1b78/generator/app.go#L453) .

Just corrected all the wrong spellings. thx

File: wg-etcd-operator/charter.md
Comment (ahrtr) [reply wg-etcd-operator/charter.md]: updated & thx

File: wg-etcd-operator/charter.md
Comment (neolit123) [inline wg-etcd-operator/charter.md:26]:
```suggestion
- Manage etcd clusters which are used as the storage backend of a host (non-nested) kube-apiservers.
```

trying to explain this better.

File: wg-etcd-operator/charter.md
Comment (neolit123) [inline wg-etcd-operator/charter.md:44]:
```suggestion
- A new repository "etcd-operator" owned by SIG etcd, and it should have implemented the very basic functionalities below:
  - Creation of a new etcd cluster with one or multiple members.
  - Scale in and out the etcd cluster.
  - Upgrading patch and minor versions.
```

**Review by neolit123 — COMMENTED** [review]
added some suggestions.

suggested to replace & with 'and' for more format language.

File: wg-etcd-operator/charter.md
Comment (ahrtr) [reply wg-etcd-operator/charter.md]: updated & thx

File: sigs.yaml
Comment (ahrtr) [reply sigs.yaml]: updated & thx.

File: wg-etcd-operator/README.md
Comment (ahrtr) [reply wg-etcd-operator/README.md]: Updated & thx

**Comment by serathius** [comment]
LGTM, awesome to see renewed interest into etcd-operator.  For me this falls into making "etcd simple to operate" as proposed by the [SIG-etcd vision](https://github.com/kubernetes/community/blob/master/sig-etcd/vision.md#etcd-is-simple-to-operate) making it a worthy investment.

**Comment by ahrtr** [comment]
/hold

I am going to send an email "WG-Creation-Request: WG etcd-operator" to [dev@kubernetes.io](dev@kubenetes.io) in the following 1~2 weeks.

**Review by jmhbnz — COMMENTED** [review]
Overall this looks good, thanks for drafting @ahrtr. A few thoughts below.

File: wg-etcd-operator/charter.md
Comment (ahrtr) [reply wg-etcd-operator/charter.md:38]: Updated & thx

File: wg-etcd-operator/charter.md
Comment (ahrtr) [reply wg-etcd-operator/charter.md]: updated & thx

File: sigs.yaml
Comment (ahrtr) [reply sigs.yaml]: updated & thx

File: sigs.yaml
Comment (ahrtr) [reply sigs.yaml]: thx

File: sig-etcd/README.md
Comment (ahrtr) [reply sig-etcd/README.md]: Updated , although `Etcd` looks more consistent to me.

File: wg-etcd-operator/charter.md
Comment (ahrtr) [reply wg-etcd-operator/charter.md]: Partially updated, thx. 

`Upgrading patch versions or one minor version` seems better, because we support upgrade across multiple patch version, but only one minor version.

File: sigs.yaml
Comment (ahrtr) [inline sigs.yaml:3418]: @jberkus could you please create a shared google doc in Kubernetes workspace, and share edit permission to both etcd-dev and cluster lifecycle?

File: OWNERS_ALIASES
Comment (ahrtr) [inline OWNERS_ALIASES:137]: @justinsb still waiting for your confirmation :)

Comment (neolit123) [reply OWNERS_ALIASES:137]: Justin was in yesterdays sig cl meeting. my suggestion is to add him only  later if he has more time.

File: sigs.yaml
Comment (neolit123) [reply sigs.yaml:3421]: up to sig etcd to decide if it would be too noisy. new channel seems better to me

File: OWNERS_ALIASES
Comment (ahrtr) [reply OWNERS_ALIASES:137]: thx @neolit123

File: sigs.yaml
Comment (neolit123) [reply sigs.yaml:3421]: IIUC, all these non existing bits like recording URL and slack can be left TBD or empty and filed later

File: OWNERS_ALIASES
Comment (ahrtr) [reply OWNERS_ALIASES]: removed for now

File: sigs.yaml
Comment (ahrtr) [reply sigs.yaml]: Applied "TBD", thx

Comment (jberkus) [reply sigs.yaml]: we can create a new slack channel if we want.

File: sigs.yaml
Comment (jberkus) [reply sigs.yaml]: what's the name/format of the doc?  Is this for meeting notes, or something else?

Comment (ahrtr) [reply sigs.yaml]: I think just we just need a google doc to meeting notes.

File: sigs.yaml
Comment (ahrtr) [reply sigs.yaml]: yes, please.

File: sigs.yaml
Comment (ahrtr) [reply sigs.yaml]: @jberkus 

Name:  `Kubernetes etcd-operator Working Group Minutes & Agenda`

We just need to follow similar format as [wg-lts](https://docs.google.com/document/d/1RI_EL35MwQxrHqlWvtQNINhOSWergL3hmOSgC5PeZss/edit#heading=h.q66rfs3opd1m)

Comment (jberkus) [reply sigs.yaml]: Bit of a holdup with google drive permissions.  I'll get this created within then next couple days.

Comment (jberkus) [reply sigs.yaml]: See request here: #7937

Comment (ahrtr) [reply sigs.yaml]: Thanks

File: wg-etcd-operator/charter.md
Comment (vincepri) [inline wg-etcd-operator/charter.md:21]: Not sure if this was already discussed, would a possible integration with Cluster API in SIG Cluster Lifecycle also be in scope? Can be as much as "it works with as an addon" or provider of sorts. It'd be great to just keep it in mind for native support of external etcd w/ kubeadm

cc @fabriziopandini @sbueringer

Comment (ahrtr) [reply wg-etcd-operator/charter.md:21]: > It'd be great to just keep it in mind for native support of external etcd w/ kubeadm

thx for the comment. It's already been discussed. It might be a potential long-term goal, but not in the scope for now. The intention is to keep it as simple as possible in the first place. It's also clarified in the section "Out of scope" (see below),


```
### Out of scope

- Manage etcd clusters running within non-Kubernetes environments.
- Manage etcd clusters which are used as the storage backend of a host (non-nested) kube-apiserver.
```

Comment (sbueringer) [reply wg-etcd-operator/charter.md:21]: I think CAPI would be doing neither of those. CAPI would be using it roughly like any other application.

We would host etcd on one Kubernetes cluster and the apiserver of another cluster would use it as backend. So no chicken-egg problems I think

File: sig-etcd/README.md
Comment (pacoxu) [inline sig-etcd/README.md:58]: Probably, we can add the etcd operator subproject info here as this is in scope of the wg.

Comment (ahrtr) [reply sig-etcd/README.md:58]: thx for the reminder. 
The subproject `etcd-io/etcd-operator` hasn't been created yet. Will update here once it's created in future.

File: wg-etcd-operator/charter.md
Comment (ahrtr) [reply wg-etcd-operator/charter.md:21]: thx for the info.

Seems like the use case is similar to the [Hosted Control Plane](https://kccnceu2024.sched.com/event/1YeQ1).

The etcd cluster, which is managed by the etcd-operator, will run as PODs within Kubernetes environment, treating them like any other typical applications. I think we are open to any use cases on top of this base.

Regarding integrating with Cluster API to support the use case of external etcd with Kubeadm, we need cluster API maintainers' help to make it work, and it's open to get it included in the roadmap (pls see the section "In Scope") of the subproject.

Comment (makhov) [reply wg-etcd-operator/charter.md:21]: I think, it's responsibility of the ControlPlane Provider to deploy etcd, not the CAPI manager itself. That's how we do in in [k0smotron](https://github.com/k0sproject/k0smotron/) ControlPlane Provider, other implementations do more or less the same.

File: OWNERS_ALIASES
Comment (justinsb) [reply OWNERS_ALIASES]: Ah sorry, I'm happy either way!  I do get way too many notifications as it is, but I am planning on participating as a contributor at least!

Comment (ahrtr) [reply OWNERS_ALIASES:137]: thx for the confirmation. Just added your back @justinsb

Comment (kvaps) [reply OWNERS_ALIASES:137]: @ahrtr we also would like to participate in etcd-operator and contribute to this project, even if it will be based on different code base.

@kvaps @sergeyshevch are Kuberentes-SIGs members

@sircthulhu @aobort @lllamnyp @Kirill-Garbar @Uburro are not, should we create PR to join them into SIG's?

Comment (ahrtr) [reply OWNERS_ALIASES:137]: @kvaps Thanks for your constant interest on this. I think anyone is welcome to participate. We definitely need community help on this project.

Comment (serathius) [reply OWNERS_ALIASES:137]: @kvaps you are free to participate, this however is not a list of members.

File: sigs.yaml
Comment (jberkus) [reply sigs.yaml]: so we'll get the channel once Steering approves the new WG.  

The channel name will be #wg-etcd-operator

**Comment by ahrtr** [comment]
> Hi! The [aenix's etcd-operator](https://github.com/aenix-io/etcd-operator) team is here.
> 
> We have [multiple regular contributors](https://github.com/aenix-io/etcd-operator/graphs/contributors) to the project and weekly meetings established. We would like to participate in this project and are open to donating our etcd-operator code here (if applicable)
> 
> We started this initiative to write a fully community-driven etcd-operator a few months ago and [informed](https://kubernetes.slack.com/archives/C3HD8ARJ5/p1712088958950639?thread_ts=1712088958.950639&cid=C3HD8ARJ5) sig-etcd about our intention. I am surprised that nobody mentioned us in this issue.
> 
> Related issues:
> 
> * [Add etcd-operator subproject to etcd-io #7796 (comment)](https://github.com/kubernetes/community/pull/7796#issue-2215254797)
> * [Donate etcd operator to sig-etcd and unify community effort etcd-io/etcd#17668](https://github.com/etcd-io/etcd/issues/17668)
> 
> How can we ensure this process will not run without our team?
> 
> /cc @sircthulhu @aobort @sergeyshevch @lllamnyp @Kirill-Garbar @Uburro

As mentioned in https://github.com/kubernetes/community/pull/7917#discussion_r1642443076, anyone is welcome to participate, we definitely need community help on this project.


Regarding the decision to pick an existing project or start from scratch, it's up to sig-etcd and sig-cluster-lifecycle leads.
Please note that even if we decide to start from scratch, we will most likely refer to some existing implementations, such as etcd-manager.

**Comment by kvaps** [comment]
@ahrtr thank you. Is there any planned schedule or meeting regarding this project?

**Comment by ahrtr** [comment]
> @ahrtr thank you. Is there any planned schedule or meeting regarding this project?

We need to wait for the Kubernetes Steering Committee to approve the WG creation request. Once we get approval, we will have regular WG Meeting, e.g bi-weekly.

**Comment by kvaps** [comment]
> We need to wait for the Kubernetes Steering Committee to approve the WG creation request. Once we get approval, we will have regular WG Meeting, e.g bi-weekly.

Got it! Thank you for the information :)

**Comment by pacoxu** [comment]
I learnt many projects are using etcd operators for performance issues. A simple solution is using a seperate etcd.
- kyverno has a new project for report server to save reports in a seperate etcd： https://github.com/kyverno/reports-server @realshuting 
- Cilium has a solution of using a seperate etcd for all CRDs:  @aanm  https://github.com/cilium/cilium-etcd-operator is archived.
- other scanerios like Events and more. @kvaps has a longer list in https://github.com/aenix-io/etcd-operator/issues/73 

Those projects' maintainers and users are potential etcd operator contributors and users. 

It would be great if they can at least response to our survey in https://forms.gle/5gBpzaxYtuQPWdBo9.

**Comment by neolit123** [comment]
@pacoxu 

> It would be great if they can at least response to our survey in https://forms.gle/5gBpzaxYtuQPWdBo9.

do you see this as a blocker for the WG etcd operator creation?

**Comment by neolit123** [comment]
steering committee met on July 3rd but the WG creation was not on the agenda.
in terms of the punchcard here:
https://github.com/kubernetes/community/blob/master/sig-wg-lifecycle.md#prerequisites-for-a-wg

we are all set with the exception of:
- steering majority / voting on this ticket
- sending PR sigs.yaml in k/community
- sending email about the sigs.yaml change

email was already sent to k-dev:
https://groups.google.com/a/kubernetes.io/g/dev/c/L1TgHxIkW_o

**Comment by kvaps** [comment]
Hey I just wanted to inform that we have weekly comunity meetings for developing [aenix-io/etcd-operator](https://github.com/aenix-io/etcd-operator), we'll have one of these tomorrow:

**When:** every Tuesday at 16:00 CET
**Where:** Google meet
**Link to join in:**
  https://meet.google.com/pib-tpfe-kuv

**Minutes:**
https://docs.google.com/document/d/1coiWlHOVfPAIyKqtfpYLhq9eeRPHCuuMxmNmVfo9-sA/edit


Everybody is free to join

File: wg-etcd-operator/charter.md
Comment (soltysh) [inline wg-etcd-operator/charter.md:11]: IIUC one of the main goals is to cover only operating etcd, but not the one backing the k8s cluster. I think it's important to make this explicit, so that users won't get confused about it. I'm aware it's mentioned as out of scope, but let's repeat it here. I remember when I first heard about it, that particular etcd was what I thought about.

Comment (soltysh) [inline wg-etcd-operator/charter.md:11]: Maybe something in the lines of:
```suggestion
The purpose of an etcd-operator is to operate automatically etcd clusters which run in the Kubernetes environment (with the exclusion of the etcd backing kubernetes cluster).
```

File: wg-etcd-operator/charter.md
Comment (soltysh) [inline wg-etcd-operator/charter.md:56]: A lot of our WGs have a hard time figuring out the exact timing. The next paragraph is sufficient to match the requirements, though. So I'd suggest dropping this statement and just go with the next one, only.

File: sigs.yaml
Comment (soltysh) [inline sigs.yaml:3426]: Feel free to put my name here.

**Comment by soltysh** [comment]
> I learnt many projects are using etcd operators for performance issues. A simple solution is using a seperate etcd.
> 
>     * kyverno has a new project for report server to save reports in a seperate etcd： https://github.com/kyverno/reports-server @realshuting
> 
>     * Cilium has a solution of using a seperate etcd for all CRDs:  @aanm  https://github.com/cilium/cilium-etcd-operator is archived.
> 
>     * other scanerios like Events and more. @kvaps has a longer list in [[epic] Move project to Kubernetes-SIGs aenix-io/etcd-operator#73](https://github.com/aenix-io/etcd-operator/issues/73)
> 
> 
> Those projects' maintainers and users are potential etcd operator contributors and users.

I believe this point:
```
Bootstrap a project "etcd-operator" owned by SIG etcd which resides in the etcd-io or kubernetes-sigs Github orgs.
  - Review existing etcd operators to see if any could be forked or referenced to advance the project.
```
can be considered as covering Paco's comment. I don't think we want to explicitly mention these projects in the charter.

**Review by soltysh — COMMENTED** [review]
Some minor nits, but mostly looks good, I'll make sure to get a final vote on this topic this week from steering.

File: wg-etcd-operator/charter.md
Comment (ahrtr) [reply wg-etcd-operator/charter.md:11]: Thanks for the suggestion. Updated, it should be clearer now. PTAL.

File: wg-etcd-operator/charter.md
Comment (ahrtr) [reply wg-etcd-operator/charter.md]: Removed

File: sigs.yaml
Comment (ahrtr) [reply sigs.yaml]: Added & thx

**Review by soltysh — APPROVED** [review]
/lgtm
from steering
/hold 
for other members to chime in

**Review by BenTheElder — COMMENTED** [review]
/lgtm
/approve
[steering]

**Review by kvaps — APPROVED** [review]

**Comment by pacoxu** [comment]
/lgtm
/approve
（steering） just want to make more cloud native contributors/maintainers/users know this new wg and participate

File: OWNERS_ALIASES
Comment (kvaps) [inline OWNERS_ALIASES:138]: If it is applicable I'd like also to be in this list as a representative of our community.

Comment (jberkus) [reply OWNERS_ALIASES:138]: Thanks for volunteering for the working group!  This WG isn't a standing body to represent all etcd operators, though.  It's a specific one-time effort to create an official Kubernetes-supported operator.  As part of that, we'll be evaluating existing operators to determine if any of them can become the core of this new operator.  That means that having the author of one such operator making that evaluation, as a WG chair, isn't a good idea.

You are very welcome to participate in the working group once it gets going, including meetings, issues, commentary, and roadmap.

File: sig-list.md
Comment (ahrtr) [reply sig-list.md:67]: We will finalize the schedule later (after this PR gets merged).

**Comment by neolit123** [comment]
/hold cancel
thank you everyone for commenting.

**Pull request merged**
